### PR TITLE
Teamcity styling

### PR DESCRIPTION
**Pull request by lukeraymonddowning**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes

This PR aims to add Pest styling to the PhpStorm `TeamCity` Logger. Here are some output examples:

Screenshot 2021-07-16 at 17 53 09

Screenshot 2021-07-16 at 17 53 50

Screenshot 2021-07-16 at 17 54 09

Screenshot 2021-07-16 at 17 55 19

Screenshot 2021-07-16 at 17 55 57

### CODE REVIEW

**Comment by lukeraymonddowning** [comment]
Multiple test cases look like this when run

Screenshot 2021-07-16 at 17 43 23

In a similar style to Pest's console runner, I opted to defer error and failure messages until the end of the output, as you can see below.

Screenshot 2021-07-16 at 17 43 57

**Comment by olivernybroe** [comment]
Looks really great! 

It's not possible for us to share the output with the normal console so we don't have to maintain the output in two places?

File: src/Logging/TeamCity.php
Comment (nunomaduro) [inline src/Logging/TeamCity.php:79]: str_starts_with is PHP 8 only?

**Review by nunomaduro — CHANGES_REQUESTED** [review]

File: src/Logging/TeamCity.php
Comment (lukeraymonddowning) [reply src/Logging/TeamCity.php:79]: Lol. My bad. Will update.

**Comment by lukeraymonddowning** [comment]
> Looks really great! 
> 
> It's not possible for us to share the output with the normal console so we don't have to maintain the output in two places? 

Hadn't really thought about that but will look at the possibility of unifying the two 👍

File: src/Logging/TeamCity.php
Comment (olivernybroe) [reply src/Logging/TeamCity.php:79]: @lukeraymonddowning would it be possible to add a test that runs the teamcity logger? :) 

It's kinda bad that our CI did not catch this.

Comment (lukeraymonddowning) [reply src/Logging/TeamCity.php:79]: This actually will succeed on <8.0 because its using a polyfill from symfony. Don't like that we're relying on that though so will alter anyway.

**Comment by lukeraymonddowning** [comment]
@olivernybroe we now defer to Collision's adapter for output. However, there seems to be a little issue with test failures showing the error output:

Screenshot 2021-07-19 at 13 44 16

I imagine this is to do with test listeners and the output helpers. Any ideas?

File: src/Logging/TeamCity.php
Comment (olivernybroe) [inline src/Logging/TeamCity.php:42]: I might be a bit blind, but what replaced the part which did this? 🤔

File: tests/Visual/JUnit.php
Comment (olivernybroe) [inline tests/Visual/JUnit.php:13]: Really nice with the tests. 

Should we also test the output? We can do that by having the junit.html file saved and checking if the result is identical with the generated one 🤔

**Review by olivernybroe — COMMENTED** [review]
Nice! 

I think the code looks great and really nice cleanup. I haven't tested it out locally yet, hopefully I get time to do that soon or someone else does it 👍

File: src/Logging/TeamCity.php
Comment (lukeraymonddowning) [reply src/Logging/TeamCity.php:42]: Basically nothing. We're firing all of the events that team city were firing manually, and now that we don't share their output at all, it became redundant. Can you see an issue with it?

File: tests/Visual/JUnit.php
Comment (lukeraymonddowning) [reply tests/Visual/JUnit.php:23]: @olivernybroe I've added a test in `SingleTestOrDirectory` that runs using JUnit logging and asserts it matches the given snapshot.

File: src/Logging/TeamCity.php
Comment (lukeraymonddowning) [reply src/Logging/TeamCity.php:42]: I've gone through and manually implemented some of the events. We have to do it manually so as not to give PHPStorm too much information (because then it writes to the console) but just enough to get it to display the correct icon. If we decorate the base `TeamCity` logger it writes out too much to the console.

**Comment by lukeraymonddowning** [comment]
@nunomaduro @olivernybroe I'm pretty happy with this PR now. Could do with somebody who has PhpStorm seeing if I've missed anything. The only quirk is on skipped tests:

Screenshot 2021-07-23 at 00 10 57

In order to get the skip indicators in PhpStorm, it has to output the "Test ignored." text. I can get it to go to the bottom using `ob_start` and `ob_flush` around the event printer, but I can't get the text to disappear completely. Thoughts?

**Comment by lukeraymonddowning** [comment]
Follow-up: We can pass a space for the message on skipped tests, which is cleaner, and combined with `ob_start` results in just a bunch of new lines at the bottom of the console output:

Screenshot 2021-07-23 at 00 18 49

**Comment by olivernybroe** [comment]
Hey @lukeraymonddowning, there is unfortunately something wrong 😢 

When I have a failing test, it does not report the error message of it as part of the specific test case.
image
image
I have to go to all output to be able to see why it failed. However before you could see this on each specific test, which is really helpful when having multiple failing tests.

**Comment by lukeraymonddowning** [comment]
Thanks @olivernybroe! That's super helpful. Do you know how we could support the full teamcity events without having it print to the console?

**Comment by olivernybroe** [comment]
@lukeraymonddowning not completely no 🙁

We might be able to do a wrapper around the out object given to collision, so we can get the output from it and wrap it in a team city format? 🤔

**Comment by nunomaduro** [comment]
@lukeraymonddowning @olivernybroe Once this is ready, fell free to merge and release.

File: src/Logging/TeamCity.php
Comment (olivernybroe) [inline src/Logging/TeamCity.php:171]:
```suggestion
        $this->write("##teamcity[{$eventName}");
```

Actually we should remove this first `\n`, we always append it, so having it both places is not needed. 

I also think we need to do this because phpstorm has the method called `serviceMessageHasNewLinePrefix` which is scheduled for removal in 2021.3.

**Comment by lukeraymonddowning** [comment]
@olivernybroe I've just gone and refactored all of the work here to clean it up nicely. Ready for another review whenever you get chance.

**Review by olivernybroe — APPROVED** [review]
I am done testing this out now and it works really well! Amazingly done ❤️

**Pull request merged**
